shadowsocks/stream.go

```diff
@@ -29,32 +29,99 @@ import (
 // payloadSizeMask is the maximum size of payload in bytes.
 const payloadSizeMask = 0x3FFF // 16*1024 - 1
 
+// SaltGenerator generates unique salts to use in Shadowsocks connections.
+type SaltGenerator interface {
+	// Returns a new salt
+	GetSalt() ([]byte, error)
+}
+
+// randomSaltGenerator generates a new random salt.
+type randomSaltGenerator struct {
+	saltSize int
+}
+
+// ServerSaltGenerator generates unique salts that are secretly marked.
+type ServerSaltGenerator struct {
+	saltSize  int
+	encrypter cipher.AEAD
+}
+
+// GetSalt outputs a random salt.
+func (sg randomSaltGenerator) GetSalt() ([]byte, error) {
+	salt := make([]byte, sg.saltSize)
+	_, err := io.ReadFull(rand.Reader, salt)
+	return salt, err
+}
+
+// Number of bytes of salt to use as a marker.
+const markLen = 4
+
+// Constant to identify this marking scheme.
+var serverIndication = []byte("outline-salt-mark")
+
+func NewServerSaltGenerator(cipher shadowaead.Cipher) (ServerSaltGenerator, error) {
+	saltSize := cipher.SaltSize()
+	zerosalt := make([]byte, saltSize)
+	encrypter, err := cipher.Encrypter(zerosalt)
```

Comment:
Is this expensive? Should we cache the salt generators for the access keys?

Reply:
It's not terribly slow (we do it for every TCP connection and every UDP packet), but it does a bunch of memory allocations.  I've arranged for the salt generators to be cached in the CipherList.